### ISSUE TITLE
Uploading images to image fields via the phone camera is broken on iOS

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#4185]
I've been owning Android phones for many years now, so I have not noticed this, but I have recently asked a colleague to confirm this for me on their iPhone, and it is indeed broken.

Can we please have a few more people test in various Apple devices and confirm?

**Comment by klonos** [comment]
I have just tested this on BrowserStack...

- iPhone X, iOS 11, Chrome: works 👍
- iPhone X, iOS 11, Safari: works 👍
- iPad Pro 11 2018, iOS 12, Chrome: works 👍
- iPad Pro 11 2018, iOS 12, Safari: works 👍

Hmm, not sure what's wrong with the people that have reported this not working for them

**Comment by klonos** [comment]
...I have also tested this by adding images in CKEditor. Also works 👍

**Comment by olafgrabienski** [comment]
I've tested uploading images in CKEditor and via the post image field on iOS.

- iPhone SE, iOS 12.4.1, Chrome and Safari: works √
- iPad Pro 11, iOS 12.4, Chrome and Safari: works √

**Side note:** Images uploaded via the image field with the iPhone didn't appear with the correct orientation, e.g. upside down. I've then enabled the setting "Enable image re-orientation" in the field settings and uploaded a new image from the iPhone, again without re-orientation success. Can someone confirm the issue? If so, we should have another look at https://github.com/backdrop/backdrop-issues/issues/2887.

**Comment by Graham-72** [comment]
@olafgrabienski Is re-orientation working when uploading in CKeditor?

**Comment by olafgrabienski** [comment]
> Is re-orientation working when uploading in CKeditor?

@Graham-72 - sorry to not have mentioned it explicitly: In CKEditor it is working.

**Comment by Graham-72** [comment]
@olafgrabienski Thanks. I don't have an iPhone but a friend emailed me a photo and it arrived with its EXIF data which gives an orientation value 'Right top', it's a portrait photo. I uploaded as an image field with re-orientation set and is has been turned by 90 degrees, so in this case it worked in Backdrop.

EXIF data shows the version of the iPhone software as 13.1.3.

Can you find out the EXIF data of the image you were testing? Maybe in Backdrop we are not correctly recognising what iOS uses to denote 'upside down'?

**Comment by Graham-72** [comment]
Another thought, which version of PHP were you using? We (Backdrop) are relying on PHP function exif_read_data and then we (Backdrop) rely on the numerical value it produces for orientation. Apparently camera manufacturers differ on this (notably Apple of course). My test was with PHP 7.1.33

**Comment by Graham-72** [comment]
A further thought! My desktop image browser (IrfanView) may be converting EXIF orientation number codes into 'Right top' etc. so maybe I am not seeing what iOS is actually producing. More tests are needed.

**Comment by olafgrabienski** [comment]
@Graham-72 Thanks for your questions! Mabe it's just an issue with my phone: When I open the image with the Mac OS preview, the info dialog says "Orientation 3", which means "180 degrees – image is upside down". So, technically it's okay to show it upside down. (Now I'm wondering why with CKEditor the image wasn't displayed upside down. Maybe because of different information about orientation? An info dialog in my alternative Mac OS X file manager says for instance "Orientation 0". I'd say, unless nobody else reports orientation issues, we shouldn't invest further time investigating it.

Apart of that, the PHP version of the website was 7.2.23. (I've used a demo sandbox from https://backdropcms.org/demo/create.)

**Comment by Graham-72** [comment]
@olafgrabienski if it receives a file with EXIF orientation value 3, the re-orientation function in Backdrop's core/includes/file.inc will rotate the image by 180 degrees (- code on line 1952 and following). I would be interested to hear anyone else's reports of problems. I feel there must be some other reason for this experience.

**Comment by olafgrabienski** [comment]
> if it receives a file with EXIF orientation value 3, the re-orientation function in Backdrop's core/includes/file.inc will rotate the image by 180 degrees

Then it worked in my case, as the image *has* been rotated (i.e. it turned upside down). When I said in https://github.com/backdrop/backdrop-issues/issues/4185#issuecomment-549131624 that it "didn't appear with the correct orientation", I didn't know of the orientation value 3. (On my phone and on my computer it's displayed 'upside up' regardless of value 3.)

**Comment by Graham-72** [comment]
Thanks for this news. I wonder whether users sometimes have problems because they do not realise that the re-orientation function has to be set by admin? Personally I think it should be on by default.

**Comment by cellear** [comment]
> I've tested uploading images in CKEditor and via the post image field on iOS.
> 
> * iPhone SE, iOS 12.4.1, Chrome and Safari: works √
> * iPad Pro 11, iOS 12.4, Chrome and Safari: works √

Wait...it _does?_ 

I've not been able to get it to work.  I guess it's a support issue to figure out what I'm doing wrong :(

**Comment by cellear** [comment]
![E11ED558-EBED-4141-8252-79D55848F2D4](https://user-images.githubusercontent.com/757296/68364874-126e8d80-00e4-11ea-886d-134046cb107f.png)
![E5BC909B-C61D-49AA-B5BD-3F3E22427A14](https://user-images.githubusercontent.com/757296/68364876-13072400-00e4-11ea-8aac-437b8cc11eb8.png)
![E2BDD11C-8E37-4BC1-99A0-EA9CAA713FC0](https://user-images.githubusercontent.com/757296/68364877-139fba80-00e4-11ea-9e88-600c959ac71c.png)

**Comment by cellear** [comment]
I don’t get a “take photo” option like Github gives me. Should I?
![FA1FB334-DF6B-4B37-AD67-2FBA0E6DD94B](https://user-images.githubusercontent.com/757296/68365058-81e47d00-00e4-11ea-8204-4b67a249a96e.png)
![94B79062-752B-41E4-9836-66A882607546](https://user-images.githubusercontent.com/757296/68365059-81e47d00-00e4-11ea-9d57-c8e9426293ce.png)

**Comment by olafgrabienski** [comment]
> Is Backdrop showing you guys this kind of interface?

Yes, it is. Tested on a vanilla demo site with iPhone SE, iOS 12.4.1, Safari:

<img src="https://user-images.githubusercontent.com/14188307/68377099-192feb80-014a-11ea-8f5b-f957b1f3471f.jpg" width="320">

**Comment by Graham-72** [comment]
And I can confirm that re-orientation is working when using an Android phone and Chrome and the option to take a photo.

@klonos we seem unable to re-create this problem.

**Comment by laryn** [comment]
Total spitball -- is it possible a secure page is required for the camera to connect to the field directly?

**Comment by jenlampton** [comment]
Aha, I thought this issue had been raised before, but all my searches turned up nothing. :( Sorry for the dupe over in https://github.com/backdrop/backdrop-issues/issues/4300.  I've closed that issue in favor of this one since this one is older. 

> I have just tested this on BrowserStack...

I don't think BrowserStack is a suitable way to test this problem since we need to access the camera on an iPhone device. I'm not sure that an file access on an emulator would work exactly the same way as file access on a device.

> Total spitball -- is it possible a secure page is required for the camera to connect to the field directly?

No, because applying the change in the PR fixes the issue for iPhone users that are having the problem.

**Comment by klonos** [comment]
@jenlampton BrowserStack has been using real devices for some time now:

https://www.browserstack.com/real-device-cloud

>*What is a Real Device Cloud?*
>
>Imagine that you have access to thousands of real desktop and mobile devices, each of which can be used for testing websites and apps in real-time, under real-user conditions. These devices are hosted on cloud-based servers, thus making them accessible online at all times. Such a testing infrastructure is called a real device cloud.

**Comment by jenlampton** [comment]
Woah.

**Comment by quicksketch** [comment]
The PR at https://github.com/backdrop/backdrop/pull/3064 seems straight forward to me. We already have this bit in file.module:

```
    // Add support for mobile camera uploads if jpg is a valid extension.
    // Browser support for HTML5 extension validation is still weak at this
    // point, add more extensive extension support in the future.
    if (in_array('jpg', $extension_array)) {
      $element['upload']['#attributes']['accept'] = '*/*,capture=camera';
    }
```

Would it work to add `'*/*,image/*,capture=camera'` as the accept attribute so this would work on file fields as well?

In any case, the current approach seems fine as-is anyway. Have we confirmed it solves the problem in the demo sandbox?

**Comment by olafgrabienski** [comment]
> Have we confirmed it solves the problem in the demo sandbox?

I've tested the sandbox, and for me it works on iPhone and iPad. However, both worked for me already before this PR -- see https://github.com/backdrop/backdrop-issues/issues/4185#issuecomment-549131624 --, so I guess more testing is needed.

**Comment by Retsell** [comment]
I’m confused. Not a programmer; just need to be able to upload photos to my sites from my iPad.  Apparently it works for some but not others.  Certainly doesn’t work for me.  I have backdrop sites on 3 different servers and none of them allow image uploads.  The “Take Photo/Photo Library/Browse” options don’t show, for image fields, or Ckeditor insert image. All works fine on a PC, but not on the iPad or iPhone, either in safari or Firefox for iOS.  If there is a fix, I’d sure like to know what it is.  This thread seemed to go off on a tangent about an image orientation issue that sounds to me like a separate issue.  Thanks for any help clearing this up!

**Comment by olafgrabienski** [comment]
@retsell Thanks for your input! It would be interesting to see if the suggested PR (pull request) fixes the issue for you. Could you give the PR a try?

To test the PR easily, go to https://github.com/backdrop/backdrop/pull/3064, scroll down a bit and look for the Website link http://3064.backdrop.backdrop.qa.backdropcms.org. This is a sandbox site which includes the changes of the PR. You'll find the login credentials of the site directly below the link.

**Comment by Retsell** [comment]
@olafgrabienski - Thanks for that helpful information. I tested on that site and find that the problem was resolved for uploading to an image field.  However, it did not fix the problem for uploading images into the body text field via the editor’s image uploader. Can the same fix be applied to the Ckeditor??

Thanks!

**Comment by olafgrabienski** [comment]
Thank you, @Retsell, for testing! I confirm that the PR works for the image field but not in CKEditor image dialog. (Tested on iPhone, iOS 13.4.) @jenlampton Can you extend your PR re CKEditor?

**Comment by klonos** [comment]
@jenlampton it seems that the default for `'#type' => 'managed_file'` when the extension is jpg (the file type that i-devices seem to be using for images taken with the camera) is to set the `accept` attribute to `*/*,capture=camera` which doesn't seem to work. It needs to be `image/*,capture=camera` instead.

Here's a PR that should fix things for CKEditor + other instances where `'#type' => 'managed_file'` is used 😉

**Comment by klonos** [comment]
Hey @Retsell 👋 ...can you please try testing the PR sandbox again? Things should now also work with the "Insert image" dialog in CKEditor. Please confirm.

**Comment by olafgrabienski** [comment]
For me it works in CKEditor now, using PR https://github.com/backdrop/backdrop/pull/3064! (iPhone, iOS 13.4)

Btw, the sandbox website has a lot of available DB updates, didn't run them.

**Comment by klonos** [comment]
So OK, it seems that there's more to this 🤔 ...although this may work now, the implementation seems to not be correct, or at least not 100% according to the current standards. This means that we may face issues in the future.

There were a couple of things that have triggered my curiosity:

- Noticing that `*/*,capture=camera` doesn't seem to work, whereas `image/*,capture=camera` does
- Noticing that adding `image/*,capture=camera` inside an `if (in_array('jpg', $extension_array))` conditional for ALL file uploads might not work as expected in all cases. You see, in my PR against @jenlampton's, I've added this in the `file.module` file, within the `file_managed_file_process()` function, which makes it apply to ALL managed file uploads. Changing `*/*` to `image/*` effectively makes it so that the accepted file type changes from allowing ANY file to allowing only images, for ALL file fields that include .jpg to the allowed list of files. In other words, this may not work as expected if .jpg has been specified _in addition_ to other non-image extensions.
- Adding `accept="image/*"` to the file field means that all image extensions are allowed, whereas the intention may be to allow only .jpg.

Anyway, my research has indicated that adding `capture=` within the `accept` attribute is not accepted/allowed in the current standard (see https://www.w3.org/TR/html51/sec-forms.html#element-attrdef-input-accept) ...it may have started like that sometime back in 2010, but it was later dropped in favor of a separate `capture` attribute (see https://w3c.github.io/html-media-capture).

So instead of this:

```php
  $element['upload']['#attributes']['accept'] = 'image/*,capture=camera';
```

the proper, standards-compliant way to do things would be something like:

```php
  $element['upload']['#attributes']['accept'] = 'image/*';
  $element['upload']['#attributes']['capture'] = 'camera';
```

Furthermore, instead of "blindly" doing something like this in `file.module`:

```php
if (in_array('jpg', $extension_array)) {
  $element['upload']['#attributes']['accept'] = 'image/*,capture=camera';
}
```

...we should be doing something along these lines:

```php
// Prepend all extensions with a dot, as it's the expected format for the
// "accept" HTML attribute.
$accept_extensions = preg_filter('/^/', '.', $extension_array);
// Prepare the "accept" string.
$accept_extensions = implode(",", $accept_extensions);
// Add allowed extensions to any already existing values.
$element['upload']['#attributes']['accept'] += $accept_extensions;
// If .jpg is among the allowed extensions for the field, then allow using the
// device camera to capture images.
if (in_array('jpg', $extension_array)) {
  $element['upload']['#attributes']['capture'] += 'camera';
}
```

What will need further testing with the above:

- is `accept="image/*" capture="camera"` the only combination that works, or does `accept=".jpg" capture="camera"` work too?

**Comment by klonos** [comment]
...I should also mention that all this is not just my OCD 😅 ...the `accept` attribute defines what file types are shown to the user when they click the upload button:

- on Windows it completely hides/disallows other file extensions/types in the Explorer popup
- on Mac it fades/disables other file extensions/types in the Finder popup

...although this does not do validation (which we do anyway on submit), it is a UX improvement, as users are not presented with disallowed extensions/types.

**Comment by klonos** [comment]
There are various resources out there, but here are the most helpful I've found:

https://www.w3schools.com/tags/att_input_accept.asp
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file
https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/accept
https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/capture
https://w3c.github.io/html-media-capture/
https://www.w3.org/TR/html51/sec-forms.html#element-attrdef-input-accept

Special kudos: https://blog.addpipe.com/correct-syntax-html-media-capture

**Comment by jenlampton** [comment]
Let's try a version using the accept attribute and see if that works on iOS.

**Comment by klonos** [comment]
@jenlampton I've filed a more comprehensive PR against your branch (also accounts for any previously set/overridden "accept" and "capture" attributes). Have a look and let me know what you think.

**Comment by herbdool** [comment]
Marking this as needs work since klonos has suggested some changes

**Comment by jenlampton** [comment]
I've merged @klonos's changes into my PR. Does anyone have an iPhone to test the updated PR?

**Comment by olafgrabienski** [comment]
I have, can test in about 2 hours.

**Comment by olafgrabienski** [comment]
@jenlampton I've had a look at the new sandbox site of [the PR](https://github.com/backdrop/backdrop/pull/3064). When I try to add an image with an iPhone, the camera is starting immediately. I can't choose between taking a photo with the camera and other options like browsing the photo library of the phone. This happens both in CKEditor and using the image field.

**Comment by klonos** [comment]
🤦 ...the spec for the `capture` attribute seems to have changed to accept only the following:

```html
<input type="file"  accept="image/*" >
<input type="file" accept="image/*" capture>
<input type="file" accept="image/*" capture="user" >
<input type="file" accept="image/*" capture="environment" >
```

...where `"user"` means the user-facing camera on mobile, and `"environment"` means the front-facing camera. From what I've read so far, specifying just `capture` w/o any value works to allow both file uploads + camera, with a prompt for the user to select.

The problem is that this:

`$element['upload']['#attributes']['capture'] = '';`

...adds the attribute like so:

```html
<input type="file" accept="image/*" capture="">
```

Which won't work. I'll need to test some more and see if `$element['upload']['#attributes']['capture'] = NULL;` works 🤔

**Comment by Retsell** [comment]
> @jenlampton I've had a look at the new sandbox site of [the PR](https://github.com/backdrop/backdrop/pull/3064). When I try to add an image with an iPhone (iOS 13.5.x), the camera is starting immediately. I can't choose between taking a photo with the camera and other options like browsing the photo library of the phone. This happens both in CKEditor and using the image field.

For me this is a big issue. The ability to upload images from an iPad (or iPhone) is critical. Maintaining a site while on the road, is severely hampered without any way to upload images. Uploading directly from the camera is much less important. I have a D7 site ready to convert to Backdrop, but can’t do it until this is fixed!

**Comment by jenlampton** [comment]
> ..adds the attribute like so: `<input type="file" accept="image/*" capture="">` ... Which won't work

@klonos Are you sure? That's valid HTML, it should work exactly the same way as `<input type="file" accept="image/*" capture>`  From @olafgrabienski 's latest test it sounds like it's

@Retsell thanks so much for weighing in. We agree this is a huge issue and that's why we got it working back in 2016, unfortunately iPhones have changed since then :) As you can see from the comments above, we've got a fix ready, and are in process of testing!

**Comment by olafgrabienski** [comment]
> From @olafgrabienski 's latest test it sounds like it's

Not sure what that means. To be clear, during my latest test I wasn't able to choose between taking a photo with the camera and browsing the phone's photo library, instead the camera started right away.

**Comment by jenlampton** [comment]
Sorry, stopped in the middle of a sentence there. I Edited the comment and Added the end :)

**Comment by Retsell** [comment]
Well here’s an interesting tidbit I discovered about this problem: The upload dialog works as it should if you remove ‘jpg’ from the list of acceptable file types in the content type > image field setting!  

‘jpeg’ is ok, as are ‘gif’ and ‘png’, but as soon as you add ‘jpg’ into the mix, the whole upload dialog box changes to the un-working version.

So a good work-around is to remove ‘jpg’ as a choice in the settings for an image field, and rename .jpg files to .jpeg before uploading. Then the files can be uploaded as usual.

Hopefully this information will also help someone to troubleshoot the issue.

**Comment by findlabnet** [comment]
Are iphone users is 80% of our users? If so, you can change the default file extension for them, but I'm not sure about this number.

**Comment by Retsell** [comment]
No, changing the default file extensions is not a solution. It’s a bug that needs to be fixed. But in the meantime, if anyone needs to be able to get image uploads working on an iPad, just remove ‘jpg’ from the allowed extensions list and it will work. Any .jpg files can be changed to .jpeg, and then they will upload.

Also, the same problem exists in file uploads — but they don’t normally include .jpg extensions. If you ad them though, it will break the uploader when used on an iOS device.

**Comment by indigoxela** [comment]
> just remove ‘jpg’ from the allowed extensions list and it will work. Any .jpg files can be changed to .jpeg, and then they will upload.

@Retsell many thanks for figuring this out! Not sure yet, how we could fix that in Backdrop, though, as it looks like an iOS bug, actually. Tricky.

**Comment by Retsell** [comment]
> Not sure yet, how we could fix that in Backdrop, though, as it looks like an iOS bug, actually. Tricky.

yes, it seems like maybe an iOS bug, but then the same file upload routine in Drupal 7 works fine with the iPad. File uploading in other systems or programs also work just fine.  The only place I’ve found this bug is in Backdrop.

**Comment by Retsell** [comment]
More about this issue: the original posting was about failure to be able to upload images from iPhone cameras. From what I discovered in testing, per my notes above:
1. When attempting to upload an image on an iOS device, the proper dialog box, which includes three options (Take Photo or Video, Photo Library, or Browse) does not appear if ‘jpg’ is included in the allowable extensions list for the image field. Instead you get a browser box with all entries grayed out.
2. Removing ‘jpg’ from the allowable extensions list causes the proper dialog box to appear again, giving the option to take a photo to upload (as well as the other two options.)
3. However, the iOS photos taken are assigned the ‘jpg’ extension. So although the proper dialog box is there, and it allows you to take a photo for uploading, it throws an error when you accept the photo, since it doesn’t have an acceptable file name extension.

Again, these functions work fine in Drupal 7, but for some reason are messed up in Backdrop.

But wait, there’s more!  What about inserting images in CKeditor?  The same problem exists there, i.e., it brings up the browser box with grayed out images instead of the three options box.  So it’s probably the same issue. But you go to the CKEditor settings for the editor format of choice, and there is no setting for allowable extensions for images. Immediately following the Image Uploading section is a File Uploading section, which DOES include settings for allowable file extensions.  Clearly there should be a field for specifying allowable extensions for images.

**Comment by Retsell** [comment]
A work-around for this issue was reported in the Forum by amilenkov. It is to install the File Resumable Upload module, 
[https://backdropcms.org/project/file_resup](url). See: [https://forum.backdropcms.org/forum/upload-images-not-working-ios-devices](url)

**Comment by klonos** [comment]
Thanks for that info @Retsell ...I still want us to fix this in core properly. It's just that last time I had a look, it wasn't as simple as it seemed, and since this requires testing in multiple devices/browsers, it'll take some time. I still want to get back to this eventually.

**Comment by Retsell** [comment]
@klonos, Yes, the issue still must be resolved. The work-around only applies to image fields, but not to image uploads using Ckeditor, or for image fields on Webforms.

**Comment by jenlampton** [comment]
If this is not a problem in D7, perhaps a quick fix would be to revert the change we added to make uploads work on mobile devices?

**Comment by Retsell** [comment]
I can confirm that in D7 image uploads work properly on iOS devices.

**Comment by laryn** [comment]
I recently got an iOS testing device so thought I'd look at this issue. I added this on the latest PR as well, but for the record:

> I've done a little spitballing and if I unset this "capture" attribute completely, iOS will then give the option to select from `Photo Library`, `Take Photo`, or `Browse`. If "capture" is there and blank, or filled with "camera" or "capture" then it goes straight to the camera.

> And I've just tested in Firefox on Android and it also seems to give the option to select from `Camera`, or `Files`.

> Do we need this "capture" attribute at all?

**Comment by laryn** [comment]
It seems that maybe "camera" is not even a valid value for this "capture" attribute!

 - https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/capture

**Comment by klonos** [comment]
Hmm 🤔 ...I'll have another look as soon as possible.

Either way, I don't think that this will be able to make it for the upcoming 1.19.4.

**Comment by Retsell** [comment]
> > Do we need this "capture" attribute at all?

No, I don't think we do.  "Take Photo" will accomplish the same thing. But "Photo Library" and "Browse" are the most important options.

**Comment by mazzech** [comment]
I just have a customer inquiry which should make camera access available for the webform data field... would this milestone also fix this?

**Comment by laryn** [comment]
I think I'm going to advocate for this issue. @jenlampton Are you interested in revising your PR or do you mind if I revise it?

**Comment by laryn** [comment]
Thanks for raising again. I may try to create an alternate PR this week and if you can test we may be able to get it in.
